File: app/webhook.py
import asyncio
import hashlib
import hmac
import json
import logging
import re
import time
from collections import OrderedDict

from fastapi import APIRouter, Request, Response

from app.config import settings
from app.remediation import trigger_remediation
from app.slack_client import get_message_attachments, get_message_text

logger = logging.getLogger(__name__)
router = APIRouter()

ROCKET_EMOJI = "rocket"

# Idempotency cache: tracks processed (channel, message_ts, user) tuples
# to prevent duplicate processing from Slack retries. LRU with max 10k entries.
_processed_reactions: OrderedDict[str, float] = OrderedDict()
_IDEMPOTENCY_MAX_SIZE = 10000
_IDEMPOTENCY_TTL_SECONDS = 300  # 5 minutes


def _is_duplicate_reaction(channel: str, message_ts: str, user: str) -> bool:
    """Check if this reaction was already processed (idempotency guard)."""
    key = f"{channel}:{message_ts}:{user}"
    now = time.time()

    # Evict expired entries periodically
    if len(_processed_reactions) > _IDEMPOTENCY_MAX_SIZE:
        # Remove oldest 20% to avoid evicting on every call
        to_remove = _IDEMPOTENCY_MAX_SIZE // 5
        for _ in range(to_remove):
            _processed_reactions.popitem(last=False)

    if key in _processed_reactions:
        # Still within TTL?
        if now - _processed_reactions[key] < _IDEMPOTENCY_TTL_SECONDS:
            return True
        # Expired — allow reprocessing
        del _processed_reactions[key]

    _processed_reactions[key] = now
    return False


def verify_slack_signature(
    body: bytes,
    timestamp: str,
    signature: str,
) -> bool:
    """Verify Slack request signature."""
    if not settings.slack_signing_secret:
        logger.warning("No Slack signing secret configured — rejecting request")
        return False

    # Check timestamp is recent (within 5 minutes)
    try:
        if abs(time.time() - int(timestamp)) > 300:
            return False
    except (ValueError, TypeError):
        return False

    sig_basestring = f"v0:{timestamp}:{body.decode('utf-8')}"
    computed = "v0=" + hmac.HMAC(
        settings.slack_signing_secret.encode(),
        sig_basestring.encode(),
        hashlib.sha256,
    ).hexdigest()

    return hmac.compare_digest(computed, signature)


def extract_github_issue_url(text: str, attachments: list[dict] | None = None) -> str | None:
    """Extract a GitHub issue URL from message text or attachments."""
    # Try message text first
    match = re.search(r"https://github\.com/[^/]+/[^/]+/issues/\d+", text or "")
    if match:
        return match.group(0)

    # Try attachments (GitHub integration uses these)
    if attachments:
        for att in attachments:
            # Check title_link
            title_link = att.get("title_link", "")
            if "issues/" in title_link:
                match = re.search(r"https://github\.com/[^/]+/[^/]+/issues/\d+", title_link)
                if match:
                    return match.group(0)
            # Check fallback text
            fallback = att.get("fallback", "")
            match = re.search(r"https://github\.com/[^/]+/[^/]+/issues/\d+", fallback)
            if match:
                return match.group(0)
            # Check text field
            att_text = att.get("text", "")
            match = re.search(r"https://github\.com/[^/]+/[^/]+/issues/\d+", att_text)
            if match:
                return match.group(0)

    return None


@router.post("/webhook/slack")
async def slack_webhook(request: Request) -> Response:
    """Handle Slack Events API callbacks. Rate-limited per IP."""
    from app.rate_limit import check_rate_limit
    await check_rate_limit(request)

    body = await request.body()

    # Verify signature first (before parsing JSON)
    timestamp = request.headers.get("x-slack-request-timestamp", "")
    signature = request.headers.get("x-slack-signature", "")
    if not verify_slack_signature(body, timestamp, signature):
        logger.warning("Invalid Slack signature")
        return Response(status_code=401)

    try:
        payload = json.loads(body)
    except (json.JSONDecodeError, ValueError):
        return Response(status_code=400)

    # Handle Slack URL verification challenge (after signature is verified)
    if payload.get("type") == "url_verification":
        return Response(
            content=payload["challenge"],
            media_type="text/plain",
        )

    # Process events
    event = payload.get("event", {})
    event_type = event.get("type")

    if event_type == "reaction_added":
        asyncio.create_task(handle_reaction_added(event))

    # Always respond 200 quickly to Slack
    return Response(status_code=200)


async def handle_reaction_added(event: dict) -> None:
    """Handle a reaction_added event — trigger remediation on :rocket:."""
    channel = ""
    message_ts = ""
    try:
        reaction = event.get("reaction", "")
        if reaction != ROCKET_EMOJI:
            return

        channel = event.get("item", {}).get("channel", "")
        message_ts = event.get("item", {}).get("ts", "")
        user = event.get("user", "unknown")

        if not channel or not message_ts:
            logger.warning("reaction_added event missing channel or ts")
            return

        # Idempotency check — prevent duplicate processing from Slack retries
        if _is_duplicate_reaction(channel, message_ts, user):
            logger.info("Duplicate reaction event (already processed): %s/%s/%s", channel, message_ts, user)
            return

        # Reject if no channel is configured (fail closed)
        if not settings.slack_channel_id:
            logger.warning("SLACK_CHANNEL_ID not configured — ignoring all reactions")
            return

        # Only process reactions from the configured channel
        if channel != settings.slack_channel_id:
            logger.info("Ignoring reaction from non-configured channel %s", channel)
            return

        logger.info(
            "Rocket reaction by %s on message %s in channel %s",
            user, message_ts, channel,
        )

        # Fetch the message to extract GitHub issue URL
        text = await get_message_text(channel, message_ts)
        attachments = await get_message_attachments(channel, message_ts)
        issue_url = extract_github_issue_url(text, attachments)

        if not issue_url:
            logger.info("No GitHub issue URL found in message %s", message_ts)
            return

        logger.info("Found issue URL: %s — triggering remediation", issue_url)

        result = await trigger_remediation(
            issue_url=issue_url,
            triggered_by=f"slack_user:{user}",
            slack_channel=channel,
            slack_message_ts=message_ts,
        )

        if not result["ok"]:
            from app.slack_client import post_thread_reply
            await post_thread_reply(
                channel,
                message_ts,
                f"⚠️ Could not start remediation: {result['error']}",
            )

    except Exception:
        logger.exception("Unhandled error in handle_reaction_added")
        if channel and message_ts:
            try:
                from app.slack_client import post_thread_reply
                await post_thread_reply(
                    channel,
                    message_ts,
                    "❌ An internal error occurred while processing this reaction. Please try again.",
                )
            except Exception:
                logger.exception("Failed to send error notification to Slack")

File: worker/src/routes/webhook.ts
import { Hono } from "hono";
import type { Env, SlackEventPayload } from "../types";
import { verifySlackSignature } from "../middleware/slack-verify";
import { checkRateLimit } from "../middleware/rate-limit";
import { checkIdempotency, findExistingActiveJob, findCompletedJobWithPr, createJob } from "../db/queries";
import { getMessage, getMessageText, getMessageAttachments, postThreadReply, getUserEmail, getUserDisplayName, isSlackBot } from "../services/slack";
import { parseIssueUrl, parsePrUrl, getIssue, findGitHubUserByEmail, approvePullRequest, assignIssue, arePrChecksPassing, mergePullRequest } from "../services/github";
import { createSession } from "../services/devin";
import { logAuditEvent } from "../db/audit";
import { enqueueDeadLetter } from "../db/dead-letters";
import { enqueuePendingMerge } from "../db/pending-merges";
import { extractGithubIssueUrl, extractGithubPrUrl } from "../services/url-extract";
import { logInfo, logError } from "../services/logger";

const ROCKET_EMOJI = "rocket";
const APPROVE_EMOJI = "white_check_mark";

const app = new Hono<{ Bindings: Env; Variables: { rawBody: string } }>();


app.post("/webhook/slack", verifySlackSignature, checkRateLimit, async (c) => {
  const body = c.get("rawBody");
  const payload: SlackEventPayload = JSON.parse(body);

  // Handle URL verification challenge
  if (payload.type === "url_verification") {
    return c.text(payload.challenge ?? "");
  }

  // Process events
  const event = payload.event;
  if (!event || event.type !== "reaction_added") {
    return c.json({ ok: true });
  }

  const channel = event.item.channel;
  const messageTs = event.item.ts;
  const user = event.user;

  // Channel restriction — applies to all reaction types
  if (c.env.SLACK_CHANNEL_ID && channel !== c.env.SLACK_CHANNEL_ID) {
    return c.json({ ok: true });
  }

  // Route by reaction type
  if (event.reaction === APPROVE_EMOJI) {
    // Idempotency: prevent duplicate approvals from Slack retries
    const approvalKey = `approve:${channel}:${messageTs}:${user}`;
    const isDuplicate = await checkIdempotency(c.env.DB, approvalKey);
    if (isDuplicate) {
      return c.json({ ok: true });
    }

    c.executionCtx.waitUntil(
      handleApproval(c.env, channel, messageTs, user)
    );
    return c.json({ ok: true });
  }

  if (event.reaction !== ROCKET_EMOJI) {
    return c.json({ ok: true });
  }

  // Reject bot users from triggering expensive Devin sessions
  const isBotUser = await isSlackBot(c.env, user);
  if (isBotUser) {
    return c.json({ ok: true });
  }

  // Idempotency check (per-user: prevents Slack retries)
  const idempotencyKey = `${channel}:${messageTs}:${user}`;
  const isDuplicate = await checkIdempotency(c.env.DB, idempotencyKey);
  if (isDuplicate) {
    return c.json({ ok: true });
  }

  // Per-message lock: prevents near-simultaneous reactions from creating
  // duplicate sessions. Short TTL (60s) so retry reactions work after failure.
  // findExistingActiveJob in handleRemediation provides long-term dedup.
  const messageLockKey = `msg_lock:${channel}:${messageTs}`;
  const isLocked = await checkIdempotency(c.env.DB, messageLockKey, 60);
  if (isLocked) {
    return c.json({ ok: true });
  }

  // Process remediation in background
  c.executionCtx.waitUntil(
    handleRemediation(c.env, channel, messageTs, user)
  );

  return c.json({ ok: true });
});

// Orchestrates the full remediation flow: fetch issue, create Devin session, notify Slack.
async function handleRemediation(
  env: Env,
  channel: string,
  messageTs: string,
  user: string
): Promise<void> {
  try {
    // Fetch message to extract issue URL (single API call)
    const msg = await getMessage(env, channel, messageTs);
    const issueUrl = extractGithubIssueUrl(msg.text, msg.attachments);

    if (!issueUrl) {
      logInfo("no_issue_url", { messageTs });
      return;
    }

    // Parse issue URL
    const parsed = parseIssueUrl(issueUrl);
    if (!parsed) {
      logInfo("unparseable_issue_url", { issueUrl });
      return;
    }

    // Restrict to configured repository
    const configuredRepo = env.GITHUB_REPO;
    if (configuredRepo && `${parsed.owner}/${parsed.repo}` !== configuredRepo) {
      logInfo("repo_mismatch", { repo: `${parsed.owner}/${parsed.repo}`, configured: configuredRepo });
      return;
    }

    // Check for existing active job
    const existingJob = await findExistingActiveJob(env.DB, issueUrl);
    if (existingJob) {
      await postThreadReply(
        env,
        channel,
        messageTs,
        `⚠️ Remediation already in progress for issue #${parsed.number}.`
      );
      return;
    }

    // Check if there's already a PR
    const completedJob = await findCompletedJobWithPr(env.DB, issueUrl);
    if (completedJob) {
      await postThreadReply(
        env,
        channel,
        messageTs,
        `✅ A PR already exists for issue #${parsed.number}: ${completedJob.pr_url}`
      );
      return;
    }

    // Fetch issue details from GitHub
    const issue = await getIssue(env, parsed.owner, parsed.repo, parsed.number);
    if (!issue) {
      await postThreadReply(
        env,
        channel,
        messageTs,
        `❌ Could not find GitHub issue #${parsed.number}.`
      );
      return;
    }

    // Only remediate open issues
    if (issue.state !== "open") {
      await postThreadReply(
        env,
        channel,
        messageTs,
        `⚠️ Issue #${parsed.number} is already ${issue.state}. Skipping remediation.`
      );
      return;
    }

    // Create Devin session
    const prompt = `Fix the following GitHub issue: ${issueUrl}\n\nTitle: ${issue.title}\n\nPlease investigate the issue, implement a fix, and create a pull request.`;
    const session = await createSession(env, prompt);

    // Store as slack_user:<id>:<display_name> for both mention and dashboard display
    let triggeredByLabel = `slack_user:${user}`;
    try {
      const name = await getUserDisplayName(env, user);
      if (name) triggeredByLabel = `slack_user:${user}:${name}`;
    } catch { /* fall back to user ID only */ }

    // Create job record
    await createJob(env.DB, {
      issue_url: issueUrl,
      issue_number: parsed.number,
      issue_title: issue.title,
      session_id: session.sessionId,
      session_url: session.url,
      triggered_by: triggeredByLabel,
      slack_channel: channel,
      slack_message_ts: messageTs,
    });

    // Assign issue to the triggering user's GitHub account (non-critical)
    let ghUsername: string | null = null;
    try {
      const email = await getUserEmail(env, user);
      if (email) {
        ghUsername = await findGitHubUserByEmail(env, email);
        if (ghUsername) {
          const assigned = await assignIssue(env, parsed.owner, parsed.repo, parsed.number, ghUsername);
          if (!assigned) ghUsername = null;
        }
      }
    } catch (assignErr) {
      logError("issue_assignment_failed", assignErr, { issueNumber: parsed.number });
    }

    // Audit log: non-critical — don't let failures mask successful remediation
    try {
      await logAuditEvent(env.DB, {
        action: "remediation_started",
        actor_slack_id: user,
        actor_github: ghUsername ?? undefined,
        target: issueUrl,
        details: `Issue #${parsed.number}: ${issue.title} → Session ${session.sessionId}`,
      });
    } catch (auditErr) {
      logError("audit_log_write_failed", auditErr);
    }

    // Notify in thread
    const mention = ghUsername ? ` (assigned to @${ghUsername})` : "";
    await postThreadReply(
      env,
      channel,
      messageTs,
      `🚀 Remediation started for issue #${parsed.number}: "${issue.title}"${mention}\n🔗 Session: ${session.url}`
    );
  } catch (err) {
    logError("handle_remediation_failed", err, { channel, messageTs });

    // Dead-letter queue: store failed event for retry
    try {
      await enqueueDeadLetter(
        env.DB,
        "remediation",
        JSON.stringify({ channel, messageTs, user }),
        err instanceof Error ? err.message : "Unknown error"
      );
    } catch (dlErr) {
      logError("dead_letter_enqueue_failed", dlErr);
    }

    await postThreadReply(
      env,
      channel,
      messageTs,
      `❌ An internal error occurred while processing this reaction. Please try again.`
    );
  }
}

// Handles ✅ reaction: maps Slack user to GitHub, submits PR approval with attribution.
async function handleApproval(
  env: Env,
  channel: string,
  messageTs: string,
  slackUserId: string
): Promise<void> {
  try {
    // Approval allowlist check
    if (env.APPROVAL_ALLOWLIST) {
      const allowlist = env.APPROVAL_ALLOWLIST.split(",").map((s) => s.trim());
      if (!allowlist.includes(slackUserId)) {
        await postThreadReply(
          env,
          channel,
          messageTs,
          "⚠️ You are not authorized to approve PRs via Slack. Contact an admin to be added to the allowlist."
        );
        try {
          await logAuditEvent(env.DB, {
            action: "approval_denied",
            actor_slack_id: slackUserId,
            target: `channel:${channel}:${messageTs}`,
            details: "User not in APPROVAL_ALLOWLIST",
          });
        } catch (auditErr) {
          logError("audit_log_write_failed", auditErr);
        }
        return;
      }
    }

    // Fetch message to extract PR URL (single API call)
    const msg = await getMessage(env, channel, messageTs);
    const prUrl = extractGithubPrUrl(msg.text, msg.attachments);

    if (!prUrl) {
      return;
    }

    const parsed = parsePrUrl(prUrl);
    if (!parsed) return;

    // Restrict approvals to configured repository
    const configuredRepo = env.GITHUB_REPO;
    if (configuredRepo && `${parsed.owner}/${parsed.repo}` !== configuredRepo) {
      logInfo("approval_repo_mismatch", { repo: `${parsed.owner}/${parsed.repo}`, configured: configuredRepo });
      return;
    }

    // Look up Slack user's email
    const email = await getUserEmail(env, slackUserId);
    if (!email) {
      await postThreadReply(
        env,
        channel,
        messageTs,
        "⚠️ Could not find your email in Slack profile. Please ensure your email is set to approve PRs."
      );
      return;
    }

    // Find corresponding GitHub user
    const ghUsername = await findGitHubUserByEmail(env, email);
    const displayName = await getUserDisplayName(env, slackUserId);

    // Build attribution message
    const attribution = ghUsername
      ? `Approved by @${ghUsername} (${displayName}) via Slack ✅ reaction`
      : `Approved by ${displayName} (${email}) via Slack ✅ reaction`;

    // Submit the approval
    const success = await approvePullRequest(
      env,
      parsed.owner,
      parsed.repo,
      parsed.number,
      attribution
    );

    if (success) {
      const ghRef = ghUsername ? `@${ghUsername}` : displayName;
      await postThreadReply(
        env,
        channel,
        messageTs,
        `✅ PR #${parsed.number} approved on GitHub (by ${ghRef})`
      );

      // Audit log: non-critical
      try {
        await logAuditEvent(env.DB, {
          action: "pr_approved",
          actor_slack_id: slackUserId,
          actor_email: email,
          actor_github: ghUsername ?? undefined,
          target: prUrl,
          details: `PR #${parsed.number} in ${parsed.owner}/${parsed.repo}`,
        });
      } catch (auditErr) {
        logError("audit_log_write_failed", auditErr);
      }

      // Auto-merge if all CI checks are passing
      try {
        const { passing, sha, merged: alreadyMerged } = await arePrChecksPassing(
          env,
          parsed.owner,
          parsed.repo,
          parsed.number
        );

        if (alreadyMerged) {
          await postThreadReply(
            env,
            channel,
            messageTs,
            `ℹ️ PR #${parsed.number} is already merged.`
          );
        } else if (passing && sha) {
          const merged = await mergePullRequest(
            env,
            parsed.owner,
            parsed.repo,
            parsed.number,
            sha
          );
          if (merged) {
            await postThreadReply(
              env,
              channel,
              messageTs,
              `🎉 PR #${parsed.number} auto-merged (all checks passed)`
            );
          } else {
            // Merge failed despite checks passing (e.g., merge conflicts)
            await enqueuePendingMerge(env.DB, {
              pr_url: prUrl,
              owner: parsed.owner,
              repo: parsed.repo,
              pr_number: parsed.number,
              slack_channel: channel,
              slack_message_ts: messageTs,
            });
            await postThreadReply(
              env,
              channel,
              messageTs,
              `⏳ PR #${parsed.number} approved — will auto-merge once all checks pass.`
            );
          }
        } else {
          // CI not yet passing — queue for poller
          await enqueuePendingMerge(env.DB, {
            pr_url: prUrl,
            owner: parsed.owner,
            repo: parsed.repo,
            pr_number: parsed.number,
            slack_channel: channel,
            slack_message_ts: messageTs,
          });
          await postThreadReply(
            env,
            channel,
            messageTs,
            `⏳ PR #${parsed.number} approved — will auto-merge once all checks pass.`
          );
        }
      } catch (mergeErr) {
        logError("auto_merge_failed", mergeErr, { prNumber: parsed.number });
        // Still queue it so the poller can retry
        try {
          await enqueuePendingMerge(env.DB, {
            pr_url: prUrl,
            owner: parsed.owner,
            repo: parsed.repo,
            pr_number: parsed.number,
            slack_channel: channel,
            slack_message_ts: messageTs,
          });
        } catch { /* ignore */ }
        await postThreadReply(
          env,
          channel,
          messageTs,
          `⏳ PR #${parsed.number} approved — will auto-merge once all checks pass.`
        );
      }
    } else {
      await postThreadReply(
        env,
        channel,
        messageTs,
        `❌ Failed to approve PR #${parsed.number}. The service token may lack write access to this repo.`
      );

      // Audit log: non-critical
      try {
        await logAuditEvent(env.DB, {
          action: "pr_approval_failed",
          actor_slack_id: slackUserId,
          actor_email: email,
          target: prUrl,
          details: "GitHub API rejected the approval request",
        });
      } catch (auditErr) {
        logError("audit_log_write_failed", auditErr);
      }
    }
  } catch (err) {
    logError("handle_approval_failed", err, { channel, messageTs });
    try {
      await postThreadReply(
        env,
        channel,
        messageTs,
        `❌ An internal error occurred while processing the approval. Please try again.`
      );
    } catch {
      // ignore post failure
    }
  }
}

export const webhookRoutes = app;
